### ISSUE TITLE
Error when inputting sorted BAM files

### ISSUE THREAD

**Issue by zkstewart** [YuSugihara/QTL-seq#50]
Hello,

I have encountered an error that appears to be caused by using sorted input BAM files.

My program command looks a little like the below.

```
qtlseq -r /home/.../genome.fasta -p /home/parent.sorted.bam \
    -b1 /home/.../sample1.sorted.bam \
    # many more files listed
    -b2 /home/.../sample2.sorted.bam \
    # many more files listed
    -n1 20 \
    -n2 28 \
    -t 12 \
    --mem 5G \
    -o qtlseq_output_comparison1
```

Full error text is below.

```
[QTL-seq:2024-11-28 17:20:26] Starting QTL-seq analysis.
[QTL-seq:2024-11-28 17:20:26] Indexing the reference FASTA.
[QTL-seq:2024-11-28 17:25:11] Reference FASTA indexing completed successfully.
[QTL-seq:2024-11-28 17:25:11] Sorting BAM files.
[QTL-seq:2024-11-28 17:25:11] !!ERROR!! samtools fixmate -m qtlseq_output_comparison1/20_bam/parent.unsorted.bam - | samtools sort -m 5G -@ 12 | samtools markdup -r - - | samtools view -b -f 2 -F 2048 -o qtlseq_output_comparison1/20_bam/parent.bam >> qtlseq_output_comparison1/log/samtools.log 2>&1

please check below:

[main_samview] fail to read the header from "-".
```

The error being presented is misleading, since it's an issue being propagated from `samtools fixmate` into the later parts of the pipe. If I just run the indicated command on its own outside of QTL-seq:

```
samtools fixmate -m qtlseq_output_comparison1/20_bam/parent.unsorted.bam - | samtools sort -m 5G -@ 12 | samtools markdup -r - - | samtools view -b -f 2 -F 2048 -o qtlseq_output_comparison1/20_bam/parent.bam >> qtlseq_output_comparison1/log/samtools.log 2>&1
```

I instead see these error messages:

```
[bam_mating_core] ERROR: Coordinate sorted, require grouped/sorted by queryname.
[W::hts_set_opt] Cannot change block size for this format
samtools sort: failed to read header from "-"
samtools markdup: error reading header

[main_samview] fail to read the header from "-".
```

So in short `samtools fixmate` isn't happy that it's being given a coordinate sorted file, and I imagine this bug is related to the commit 08150e61f2b31203bcc40fba164c3eca2390f3e6 where these changes were made.

It's possible that QTL-seq needs to recognise if the input is coordinate sorted in which case `samtools fixmate` should be omitted?

**Comment by carrot881216** [comment]
hello,I have also encountered the same problem. How did you solve it

**Comment by zkstewart** [comment]
I downgraded qtlseq to version 2.2.2 with conda like:

`conda install bioconda::qtlseq=2.2.2`

From briefly looking through the commit history of QTL-seq, version 2.2.6 might be the last version where this `samtools fixmate` issue doesn't occur, but I didn't test that.

However, version 2.2.2 appears to work correctly, so you might have luck with that.

**Comment by YuSugihara** [comment]
Thank you very much for raising the error.

This is indeed a bug, and it should be fixed in the latest version v2.2.9 by adding `samtools collate` before `samtools fixmate`.

https://github.com/YuSugihara/QTL-seq/commit/c96b2b5b878ca643db2fde05a3f55a7967b5135e#diff-1986cb4c9dcb89968f7245b0f3075c5cbbb3bfb1d2feb60909217ec5be1f296f

When I input FASTQ files, this error did not occur.
This error happened when I input BAM files.
